### ISSUE TITLE
link-to + model + query-params configured with refreshModel: true causes error

### ISSUE THREAD

**Issue by mehulkar** [emberjs/ember.js#16921]
```hbs
{{#link-to 'post' model (query-params something="1")}}
Click
{{/link-to}}
```

throws an error

```
Uncaught Error: You didn't provide enough string/numeric parameters to satisfy all of the dynamic segments for route post
```

Repro:

https://ember-twiddle.com/be857bace6d8c03ff44e6e09d4111c64

**Comment by mariokostelac** [comment]
I can confirm this is still a problem. I spent some time digging and narrowed down the part causing the bug.

When query params change, https://github.com/emberjs/ember.js/blob/707a938e536cb921dbe80ee270c431d1b9717e38/packages/%40ember/-internals/routing/lib/system/route.ts#L2423 fires because `refreshModel` is set to true. That calls `router.refresh` (that calls refresh again) and we end up in https://github.com/tildeio/router.js/blob/bedbf4e2c6a544bc8a678d286d00b412d0ebab51/lib/router/router.ts#L894-L900, creating new intent, but without the context. It's interesting that there was a comment about gathering contexts https://github.com/tildeio/router.js/blob/bfcb66f7cbd52df698963ec45d07e56bf2b2de14/lib/router/router.js#L217, lost during translation to Typescript.

I would say that it would be possible to hack context from `previousTransition.intent` the same way https://github.com/tildeio/router.js/blob/c315882045f2cf5e6fa7629572cbe1c4a944782b/lib/router/router.ts#L905-L907 was done.

I suspect there are more bugs around creating the refresh intent (will try to create repro and open an issue), causing queryParams not being sticky.

**Comment by neeleshsaxena** [comment]
@mariokostelac Could you take a peek at this issue and see if it's a similar issue. It does have a repro along with it. Thanks.
https://discuss.emberjs.com/t/ember-routes-model-hook-doesnt-re-fired-when-qp-change/16361/7?u=neeleshsaxena

**Comment by mariokostelac** [comment]
@neeleshsaxena will take a look. I'll probably spend some time next week with @GavinJoyce to create at least a failing test.

**Comment by mariokostelac** [comment]
I wrote a failing test #17831.

**Comment by GavinJoyce** [comment]
Screenshot of @mariokostelac's failing test:

<img width="1180" alt="Screen Shot 2019-04-01 at 15 54 16" src="https://user-images.githubusercontent.com/2526/55337553-d19bff80-5496-11e9-9065-8ab50c3e5d63.png">

**Comment by GavinJoyce** [comment]
Poking around, I managed to get the test passing by changing the following in `router_js`. I'm unclear yet if this is a good approach:

https://github.com/tildeio/router.js/blob/726513e66fa4690af3a81f1cae8d312b1eb10340/lib/router/transition-intent/named-transition-intent.ts#L212

```js
let oldParams =
        (oldHandlerInfo && name === oldHandlerInfo.name && oldHandlerInfo.params) || {};
```

to

```js
let oldParams =
        (oldHandlerInfo && name === oldHandlerInfo.name && oldHandlerInfo.serialize()) || {};
```

**Comment by GavinJoyce** [comment]
@rwjblue do you think the change above is a reasonable approach?

**Comment by neeleshsaxena** [comment]
https://github.com/tildeio/router.js/pull/284 seems to have fixed my issue here - https://github.com/tildeio/router.js/issues/285
This isn't exactly the same issue, but atleast now refresh isn't being called again n again.
Might help updating to ember 3.9?

**Comment by matixmatix** [comment]
Is this still an issue? I'm experiencing this issue in older Ember (2.12) and wondering if this was ever resolved?

**Comment by mehulkar** [comment]
@matixmatix yes, I think still an issue. You can check the twiddle at the top and repro with more recent versions of Ember

**Comment by JackHowa** [comment]
is there any update on this issue? Experiencing it with `ember-cli@3.28.6`
